[wasm] renames and cleanup before modularization

pavelsavara opened the pull request:
- rename main javaScript files to `main.js` and `test-main.js`
- sample and test script cleanup
- rename set_exit_code method
- rewrite test start as async method
- improve script loading via script element on page in test
- use BINDING.bind_static_method instead of INTERNAL where possible
- better .d.ts exports
- formatted html files
- moved dotnet IFFE out of emcc Closure compiled section and re-enabled terser() and ES2018 target.

pavelsavara committed:
- no imports from outer scope
- move dotnet to -extern-pre-js
- re-enable JS minification
- rename main javaScript files to main.js and test-main.js
- sample and test script cleanup
- rename set_exit_code method
- rewrite test start as async method
- improve script loading via script element on page in test
- use BINDING.bind_static_method instead of INTERNAL where possible
- better .d.ts exports
- formatted html files

kg commented:
Happy to finally rename runtime to something reasonable

pavelsavara commented:
This passed CI tests with full matrix. I also tested debugger unit tests and debugger in Blazor app manually.

kg commented on src/mono/wasm/runtime/export-types.ts:
```diff
@@ -0,0 +1,7 @@
+import { EmscriptenModuleConfig } from "./types";
+
+// this is the only public export from the dotnet.js module
+declare function createDotnetRuntime(moduleFactory: (api: DotNetExports) => EmscriptenModuleConfig): Promise<DotNetExports>;
```
Maybe DotNetExports is the wrong name for this if 'createDotnetRuntime' is the only public export from dotnet.js? If the parameter name is 'api' maybe it's DotNetPublicAPI or something

kg commented on src/mono/wasm/runtime/exports.ts:
```diff
@@ -115,23 +115,48 @@ export const BINDING: BINDING = <any>{
 // it exports methods to global objects MONO, BINDING and Module in backward compatible way
 // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
-function export_to_emscripten(dotnet: any, mono: any, binding: any, internal: any, module: any): void {
+function exportAPI(mono: any, binding: any, internal: any, module: any, isGlobal: boolean, isNode: boolean, isShell: boolean, isWeb: boolean, glocateFile: Function): void {
```
Can we make all these bools and glocateFile live in an object with keys?

pavelsavara replied:
They are not bools, `mono` is seed object what would become `MONO` later. `isNode`, `locateFile` etc is import from emscripten which is not expoted on `Module` object. I will rename it to make it clearer.

kg commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -223,6 +238,20 @@ function _finalize_startup(args: MonoConfig, ctx: MonoInitContext) {
+            Module.printErr("MONO_WASM: loaded_cb () failed: " + err);
+            Module.printErr("MONO_WASM: Stacktrace: \n");
+            Module.printErr(err.stack);
+            throw err;
```
shouldn't we reject here?

kg commented on src/mono/wasm/test-main.js:
```diff
@@ -101,18 +101,12 @@ if (typeof globalThis.performance === 'undefined') {
         if (!Module.config) {
-            console.error("Could not find ./mono-config.json. Cancelling run");
-            fail_exec(1);
+            set_exit_code(1, "Could not find ./mono-config.json. Cancelling run");
```
while i think the extra code below running is harmless, we should probably 'return' here after setting the exit code anyway

kg commented on src/mono/wasm/test-main.js:
```diff
@@ -333,7 +326,17 @@ async function loadDotnet(file) {
-            return globalThis.Module;
+            let timeout = 100;
+            // bysy spin waiting for script to load into global namespace
+            while (timeout > 0) {
```
if the script being loaded has to do some initialization work, the 'load' event might be too early

maraf replied:
Ook 👍

pavelsavara replied:
with ES6, this will do away and we will only `import`.

pavelsavara commented on src/mono/wasm/runtime/exports.ts:
```diff
@@ -314,4 +337,9 @@ export interface BINDING {
+    MONO: MONO,
+    BINDING: BINDING,
+    Module: any
 }
```
Yes, this is only type export, not actual API instance. I don't want to document that we have INTERNAL in the `.d.ts` file

pavelsavara committed:
feedback from @kg and @maraf
- renamed modules.ts to imports.ts which makes more sense
- improved error propagation
- renamed __initializeImportsAndExports

pavelsavara committed:
observe exit behavior

thaystg commented:
Debugger tests -> passing.
Debug a Blazor App -> working.

pavelsavara committed:
early console redirect

pavelsavara committed:
move log about exit

pavelsavara committed:
delayed exit

pavelsavara committed:
fix build

pavelsavara committed:
Merge branch 'main' into wasm_modularization8

# Conflicts:
#	src/mono/wasm/runtime/startup.ts

pavelsavara committed:
fix

pavelsavara committed:
defensive exit
